vochain/state_snapshot.go

```diff
@@ -0,0 +1,416 @@
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/vocdoni/arbo"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/statedb"
+)
+
+const (
+	snapshotHeaderVersion = 1
+	snapshotHeaderLenSize = 32
+)
+
+// A StateSnapshot is a copy in a specific point in time of the blockchain state.
+// The state is supposed to be a list of nested merkle trees.
+// The StateSnapshot contains the methods for building a single file snapshot of
+// the state containing multiple trees.
+// The implementation allows the encoding and decoding of snapshots.
+//
+// The structure of the snapshot encoded file is:
+//
+//  [headerLen][header][tree1][tree2][treeN]
+//
+// - headerlen is a fixed 32 bytes little endian number indicating the size of the header.
+//
+// - header is the Gob encoded structure containing the information of the trees (size, name, etc.).
+//
+// - treeN is the raw bytes dump of all trees.
+type StateSnapshot struct {
+	path                  string
+	file                  *os.File
+	lock                  sync.Mutex
+	header                SnapshotHeader
+	headerSize            uint32
+	currentTree           int       // indicates the current tree
+	currentTreeReadBuffer io.Reader // the buffer for reading the current tree
+}
+
+// SnapshotHeader is the header structure of StateSnapshot containing the list of merkle trees.
+type SnapshotHeader struct {
+	Version int
+	Root    []byte
+	ChainID string
+	Height  uint32
+	Trees   []SnapshotHeaderTree
+}
+
+// SnapshotHeaderTree represents a merkle tree of the StateSnapshot.
+type SnapshotHeaderTree struct {
+	Name   string
+	Size   uint32
+	Parent string
+	Root   []byte
+}
+
+// SetMainRoot sets the root for the mail merkle tree of the state.
+func (s *StateSnapshot) SetMainRoot(root []byte) {
+	s.header.Root = root
+}
+
+// SetHeight sets the blockchain height for the snapshot.
+func (s *StateSnapshot) SetHeight(height uint32) {
+	s.header.Height = height
+}
+
+// SetChainID sets the blockchain identifier for the snapshot.
+func (s *StateSnapshot) SetChainID(chainID string) {
+	s.header.ChainID = chainID
+}
+
+// Open reads an existing snapshot file and decodes the header.
+// After calling this method everything is ready for reading the first
+// merkle tree. No need to execute `FetchNextTree` until io.EOF is reached.
+//
+// This method performs the opposite operation of `Create`, one of both needs
+// to be called (but not both).
+func (s *StateSnapshot) Open(filePath string) error {
+	var err error
+	s.path = filePath
+	s.file, err = os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	headerSizeBytes := make([]byte, snapshotHeaderLenSize)
+	if _, err = io.ReadFull(s.file, headerSizeBytes); err != nil {
+		return err
+	}
+	s.headerSize = binary.LittleEndian.Uint32(headerSizeBytes)
+	headerBytes := make([]byte, int(s.headerSize))
+	s.file.Seek(snapshotHeaderLenSize, 0)
+	if _, err = io.ReadFull(s.file, headerBytes); err != nil {
+		return err
+	}
+	decoder := gob.NewDecoder(bytes.NewBuffer(headerBytes))
```

Comment:
this can still be a LimitReader, no need for a bytes.Buffer. I think I even showed an example in my previous review? :)

Reply:
I tried and something failed (gob could not decode it). I'm not sure why, just reverted to the buffer (which works and is also OK).